plugins/renepay/mcf.c

```diff
@@ -256,16 +256,54 @@ static const s64 MU_MAX = 128;
  * I hope this will clarify my future self when I forget.
  *
  * */
-typedef union
-{
-	struct{
-		u32 dual: 1;
-		u32 part: PARTS_BITS;
-		u32 chandir: 1;
-		u32 chanidx: (32-1-PARTS_BITS-1);
-	};
+
+/*
+ * We want to use the whole number here for convenience, but
+ * we can't us a union, since bit order is implementation-defined and
+ * we want chanidx on the highest bits:
+ *
+ * [ 0       1 2     3            4 5 6 ... 31 ]
+ *   dual    part    chandir      chanidx
+ */
+struct arc {
 	u32 idx;
-} arc_t;
+};
+
+#define ARC_DUAL_BITOFF (0)
+#define ARC_PART_BITOFF (1)
+#define ARC_CHANDIR_BITOFF (1 + PARTS_BITS)
+#define ARC_CHANIDX_BITOFF (1 + PARTS_BITS + 1)
+#define ARC_CHANIDX_BITS  (32 - ARC_CHANIDX_BITOFF)
+
```

Comment:
At this point I think you should remove line 194 to avoid duplication
```
static const size_t ARC_ADDITIONAL_BITS = PARTS_BITS + 2;
```
and put here instead
```
#define ARC_ADDITIONAL_BITS (ARC_CHANIDX_BITOFF)
```
which is the same thing. This number is used to compute the `max_num_arcs` later in line 578.

Reply:
OK, I did this a bit differently: what you really want to know is "how many arcs can we have per channel", which is actually independent of the bit representation (though they're derived from a common source).  So I answered this directly in the final commit!